File: app/src/main/java/dev/ragnarok/fenrir/db/interfaces/ICommentsStorage.java
package dev.ragnarok.fenrir.db.interfaces;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;

import java.util.List;

import dev.ragnarok.fenrir.db.model.entity.CommentEntity;
import dev.ragnarok.fenrir.db.model.entity.OwnerEntities;
import dev.ragnarok.fenrir.model.CommentUpdate;
import dev.ragnarok.fenrir.model.Commented;
import dev.ragnarok.fenrir.model.DraftComment;
import dev.ragnarok.fenrir.model.criteria.CommentsCriteria;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public interface ICommentsStorage extends IStorage {

    Single<int[]> insert(int accountId, int sourceId, int sourceOwnerId, int sourceType, List<CommentEntity> dbos, OwnerEntities owners, boolean clearBefore);

    Single<List<CommentEntity>> getDbosByCriteria(@NonNull CommentsCriteria criteria);

    @CheckResult
    Maybe<DraftComment> findEditingComment(int accountId, @NonNull Commented commented);

    @CheckResult
    Single<Integer> saveDraftComment(int accountId, Commented commented, String text, int replyToUser, int replyToComment);

    Completable commitMinorUpdate(CommentUpdate update);

    Observable<CommentUpdate> observeMinorUpdates();

    Completable deleteByDbid(int accountId, Integer dbid);
}

File: app/src/main/java/dev/ragnarok/fenrir/db/interfaces/ICommentsStorage.kt
package dev.ragnarok.fenrir.db.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.db.model.entity.CommentEntity
import dev.ragnarok.fenrir.db.model.entity.OwnerEntities
import dev.ragnarok.fenrir.model.CommentUpdate
import dev.ragnarok.fenrir.model.Commented
import dev.ragnarok.fenrir.model.DraftComment
import dev.ragnarok.fenrir.model.criteria.CommentsCriteria
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single

interface ICommentsStorage : IStorage {
    fun insert(
        accountId: Int,
        sourceId: Int,
        sourceOwnerId: Int,
        sourceType: Int,
        dbos: List<CommentEntity>,
        owners: OwnerEntities?,
        clearBefore: Boolean
    ): Single<IntArray>

    fun getDbosByCriteria(criteria: CommentsCriteria): Single<List<CommentEntity>>

    @CheckResult
    fun findEditingComment(accountId: Int, commented: Commented): Maybe<DraftComment>?

    @CheckResult
    fun saveDraftComment(
        accountId: Int,
        commented: Commented,
        text: String?,
        replyToUser: Int,
        replyToComment: Int
    ): Single<Int>

    fun commitMinorUpdate(update: CommentUpdate): Completable
    fun observeMinorUpdates(): Observable<CommentUpdate>
    fun deleteByDbid(accountId: Int, dbid: Int): Completable
}